builtin/logical/database/rotation.go

```
@@ -0,0 +1,380 @@
+		// be rotated
+		if time.Now().Unix() < item.Priority {
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+			// break out of the for loop
+			break
+		}
+
+		input := &setStaticAccountInput{
+			RoleName: item.Key,
+			Role:     role,
+		}
+
+		// If there is a WAL entry related to this Role, the corresponding WAL ID
+		// should be stored in the Item's Value field.
+		if walID, ok := item.Value.(string); ok {
+			walEntry := b.findStaticWAL(ctx, s, walID)
+			if walEntry != nil && walEntry.NewPassword != "" {
+				input.Password = walEntry.NewPassword
+				input.WALID = walID
+			}
+		}
+
+		resp, err := b.setStaticAccount(ctx, s, input)
+		if err != nil {
+			b.logger.Warn("unable to rotate credentials in periodic function", "error", err)
+			// Increment the priority enough so that the next call to this method
+			// likely will not attempt to rotate it, as a back-off of sorts
+			item.Priority = time.Now().Add(10 * time.Second).Unix()
+
+			// preserve the WALID if it was returned
+			if resp != nil && resp.WALID != "" {
+				item.Value = resp.WALID
+			}
+
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+			// go to next item
+			continue
+		}
+
+		lvr := resp.RotationTime
+		if lvr.IsZero() {
+			lvr = time.Now()
+		}
+
+		// update priority and push updated Item to the queue
+		nextRotation := lvr.Add(role.StaticAccount.RotationPeriod)
+		item.Priority = nextRotation.Unix()
+		if err := b.pushItem(item); err != nil {
+			b.logger.Warn("unable to push item on to queue", "error", err)
+		}
+	}
+	return nil
+}
+
+// findStaticWAL loads a WAL entry by ID. If found, only return the WAL if it
+// is of type staticWALKey, otherwise return nil
+func (b *databaseBackend) findStaticWAL(ctx context.Context, s logical.Storage, id string) *setCredentialsWAL {
```

This function should return an error

---

Done in 97d562bfe